sample_metadata/parser/generic_parser.py

```diff
@@ -310,64 +402,133 @@ async def process_group(
+                }
+                if not args['id']:
+                    del args['id']
+
+                sequences_to_upsert.append(SequenceUpsert(**args))
 
         # Should we add or update sample
         args = {
             'id': cpg_sample_id,
-            'meta': collapsed_sample_meta,
+            'meta': collapsed_sample_meta.meta,
             'external_id': external_sample_id,
-            'type': SampleType(sample_type),
-            'sequences': [sequence_to_upsert] if sequence_to_upsert else [],
+            'type': self.get_sample_type(rows),
+            'sequences': sequences_to_upsert,
         }
-
         if not cpg_sample_id:
             del args['id']
 
-        sample_to_upsert = SampleBatchUpsertItem(**args)
+        sample_to_upsert = SampleBatchUpsert(**args)
 
         if collapsed_analyses:
-            analysis_to_add.extend(collapsed_analyses)
+            analyses_to_add.extend(collapsed_analyses)
 
         if collapsed_qc:
-            analysis_to_add.append(
+            analyses_to_add.append(
                 AnalysisModel(
                     sample_ids=['<none>'],
-                    type=AnalysisType('qc'),
-                    status=AnalysisStatus('completed'),
+                    type=self.get_analysis_type(external_sample_id, rows),
+                    status=self.get_analysis_status(external_sample_id, rows),
                     meta=collapsed_qc,
                 )
             )
 
         return (
             sample_to_upsert,
-            sequence_to_upsert,
-            analysis_to_add,
+            sequences_to_upsert,
+            analyses_to_add,
+        )
+
+    async def process_participant_group(
+        self, participant_id: int, sample_map: Dict[str, Any]
+    ):
+        """
+        ASYNC function that (maps) transforms one GroupedRow, and returns a Tuple of:
+            (
+                participants_to_upsert,
+                samples_to_upsert,
+                sequences_to_upsert,
+                analyses_to_add,
+            )
+
+        Then the calling function does the (reduce).
+        """
+
+        all_rows = [r for row in sample_map.values() for r in row]
+
+        # now we have sample / sequencing meta across 4 different rows, so collapse them
+        collapsed_participant_meta = await self.get_participant_meta(
+            participant_id, all_rows
+        )
+
+        # Get external sid to cpg map
+        existing_external_id_to_cpgid = (
+            await SampleApi().get_sample_id_map_by_external_async(
+                self.sample_metadata_project,
+                list(sample_map.keys()),
+                allow_missing=True,
+            )
+        )
+
+        # Get all the samples and sequences to upsert first
+        samples_to_upsert = []
+        sequences_to_upsert = []
+        analyses_to_add = []
+
+        for sample_id, rows in sample_map.items():
+            cpg_id = existing_external_id_to_cpgid[sample_id]
```

Comment:
They shouldn't always have a cpg_id right? I haven't run it, but I thought the direct access for `sample_id` would fail if the sample doesn't exist.

Reply:
Correct, but the process_sample_group was doing api calls to get the cpg_id. So I did one batch at the beginning and pass in both the sample_id (always present) and cpg_id (if it exists). Otherwise the cpg_id will just be None.